src/OpenTelemetry.Contrib.Instrumentation.Remoting/Implementation/TelemetryDynamicSink.cs

```diff
@@ -0,0 +1,127 @@
+﻿// <copyright file="TelemetryDynamicSink.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.Remoting.Contexts;
+using System.Runtime.Remoting.Messaging;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Remoting.Implementation
+{
+    /// <summary>
+    /// A dynamic remoting sink that intercepts all calls leaving or entering an AppDomain, see
+    /// <see cref="TracerProviderBuilderExtensions.AddRemotingInstrumentation"/> RegisterDynamicProperty
+    /// code.
+    /// </summary>
+    /// <remarks>
+    /// See https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/tdzwhfy3(v=vs.100) for
+    /// information about remoting sinks and https://docs.microsoft.com/en-us/dotnet/api/system.runtime.remoting.contexts.context.registerdynamicproperty?view=netframework-4.8
+    /// for RegisterDynamicProperty behavior.
+    /// </remarks>
+    internal class TelemetryDynamicSink : IDynamicMessageSink
+    {
+        internal const string ActivitySourceName = "OpenTelemetry.Remoting";
+        private const string ActivityOutName = ActivitySourceName + ".RequestOut";
+        private const string ActivityInName = ActivitySourceName + ".RequestIn";
+
+        // TODO: make configurable?
+        private static readonly IPropagator Propagator = new TextMapPropagator();
+
+        private static readonly Version Version = typeof(TelemetryDynamicSink).Assembly.GetName().Version;
+        private static readonly ActivitySource RemotingActivitySource = new ActivitySource(ActivitySourceName, Version.ToString());
+
+        public void ProcessMessageStart(IMessage reqMsg, bool bCliSide, bool bAsync)
+        {
+            // Are we executing on client?
+            if (bCliSide)
+            {
+                // Start new outgoing activity
+                var act = RemotingActivitySource.StartActivity(ActivityOutName, ActivityKind.Client);
+                if (act != null)
+                {
+                    act.SetTag("uri", reqMsg.Properties["__Uri"]);
```

Comment:
its recommended to protect these `SetTag` inside If (act.IsAllDataRequested) {}

Reply:
Fixed. I've changed tags (and `DisplayName`) to now be set only if `IsAllDataRequested`. I've also added some exception-specific tags which are always set in case of an exception regardless of `IsAllDataRequested`.